pages/k8s/keepalived.md

```diff
@@ -0,0 +1,85 @@
+---
+wrapper_template: "base-docs.html"
+markdown_includes:
+  nav: "shared/_side-navigation.md"
+context:
+  title: "HA for kubeapi-load-balancer"
+  description: Using keepalived to provide failover for the kube-api-loadbalancer .
+keywords: failover, ha, kube-api-loadbalancer, loadbalancer, keepalived
+tags: [operating]
+sidebar: k8smain-sidebar
+permalink: keepalived.html
+layout: [base, ubuntu-com]
+toc: False
+---
+
+The standard deployment of  the
+**Canonical Distribution of Kubernetes<sup>&reg;</sup> (CDK)** includes a single
+instance of the kube-api-loadbalancer.  For many use cases this is perfectly adequate,
+but in a production environment you should be keen to eliminate any single point of
+failure.
+
+The recommended way to provide a failover for the kube-api-loadbalancer is by using
+[keepalived][keepalived-home]. This is available as a **Juju** charm and can be deployed
+into your **CDK** model and configured as follows:
+
+1. Deploy the `keepalived` charm:
+    ```bash
+    juju deploy cs:~containers/keepalived-3
+    ```
+
+1. Add the required relations :   
+    ```bash
+    juju add-relation keepalived:juju-info kubeapi-load-balancer:juju-info
+    juju add-relation keepalived:lb-sink kubeapi-load-balancer:website
+    juju add-relation keepalived:loadbalancer kubernetes-master:loadbalancer
+    juju add-relation keepalived:website kubernetes-worker:kube-api-endpoint
+    ```
+    This redirects both the Kubernetes master and worker units to point at the keepalive
+    service reather than the api-endpoint directly.
+
+1. Configure the keepalived application. You should substitute a suitable IP address and
+     FQDN in the example below:
+    ```bash
+    export VIP_HOSTNAME=test.example.com
+    juju config keepalived virtual_ip=10.10.74.250
+    juju config keepalived port=443
+    juju config keepalived vip_hostname=$VIP_HOSTNAME
+    ```
+
+1.  Add the new hostname to the API server certificate. This is done by specifying an
+    additional [SAN][]:
+    ```bash
+    juju config kubeapi-load-balancer extra_sans=$VIP_HOSTNAME
+    juju config kubernetes-master extra_sans=$VIP_HOSTNAME
+    ```
+
+1. Wait for the new service to settle. You can check the status of the `keepalived`
```

Comment:
This, and the order of operations is slightly different to the original. In a number of runthroughs I found I got errors from workers and/or master units if the loadbalancer relation was removed before the replacement was configured and running. May have been unfortunate timing, but this seems like a more robust sequence to me